sdk/cosmos/azure-cosmos/azure/cosmos/_execution_context/aio/non_streaming_order_by_aggregator.py

```diff
@@ -0,0 +1,170 @@
+    """This class is a subclass of the query execution context base and serves for
+    non-streaming order by queries. It is very similar to the existing MultiExecutionContextAggregator,
+    but is needed since we're dealing with items and not document producers.
+
+    This class builds upon the multi-execution aggregator, building a document producer per partition
+    and draining their results entirely in order to create the result set relevant to the filters passed
+    by the user.
+    """
+
+    def __init__(self, client, resource_link, query, options, partitioned_query_ex_info):
+        super(_NonStreamingOrderByContextAggregator, self).__init__(client, options)
+
+        # use the routing provider in the client
+        self._routing_provider = client._routing_map_provider
+        self._client = client
+        self._resource_link = resource_link
+        self._query = query
+        self._partitioned_query_ex_info = partitioned_query_ex_info
+        self._sort_orders = partitioned_query_ex_info.get_order_by()
+
+        pq_size = partitioned_query_ex_info.get_top() or partitioned_query_ex_info.get_limit()
+        self._orderByPQ = FixedSizePriorityQueue(pq_size)
+
+    async def __anext__(self):
+        """Returns the next result
+
+        :return: The next result.
+        :rtype: dict
+        :raises StopIteration: If no more result is left.
+        """
+        if self._orderByPQ.size() > 0:
+            res = await self._orderByPQ.pop_async(self._document_producer_comparator)
+            return res
+        raise StopAsyncIteration
+
+    async def fetch_next_block(self):
+
+        raise NotImplementedError("You should use pipeline's fetch_next_block.")
+
+    async def _repair_document_producer(self):
+        """Repairs the document producer context by using the re-initialized routing map provider in the client,
+        which loads in a refreshed partition key range cache to re-create the partition key ranges.
+        After loading this new cache, the document producers get re-created with the new valid ranges.
+        """
+        # refresh the routing provider to get the newly initialized one post-refresh
+        self._routing_provider = self._client._routing_map_provider
+        # will be a list of (partition_min, partition_max) tuples
+        targetPartitionRanges = await self._get_target_partition_key_range()
+
+        targetPartitionQueryExecutionContextList = []
+        for partitionTargetRange in targetPartitionRanges:
+            # create and add the child execution context for the target range
+            targetPartitionQueryExecutionContextList.append(
+                self._createTargetPartitionQueryExecutionContext(partitionTargetRange)
+            )
+
+        self._doc_producers = []
+        for targetQueryExContext in targetPartitionQueryExecutionContextList:
+            try:
+                await targetQueryExContext.peek()
```

Comment:
Is peek actually pushing forward the iteration such that `StopIteration` is a possibility? e.g. the `peek` implementation above would only ever raise an `IndexError`? Or is this a different `peek`?

Reply:
yep in this case peek can actually push forward the iteration if the result is None (which could cause the StopIteration): 
![image](https://github.com/Azure/azure-sdk-for-python/assets/30335873/279bf9a9-77c0-45ce-93a9-0050951bb804)